File: ruby/lib/queue.rb
require 'mongo'

module Karait
  class Queue
    
    include Karait
    
    MESSAGES_READ = 10
    NO_OBJECT_FOUND_ERROR = 'No matching object found'
    
    def initialize(opts={})
      set_instance_variables opts
      create_mongo_connection
    end
    
    def write(message, opts={})
      if message.class == Hash
        message_dict = message
      else
        message_dict = message.to_hash
      end
      
      message_dict[:_meta] = {
        :expire => opts.fetch(:expire, -1.0),
        :timestamp => Time.now().to_f,
        :expired => false,
        :visible_after => -1.0
      }
      
      message_dict[:_meta][:routing_key] = opts.fetch(:routing_key) if opts[:routing_key]
      
      @queue_collection.insert(message_dict, :safe => true)
    end
    
    def read(opts={})
      opts = {
        :messages_read => 10,
        :visibility_timeout => -1.0,
        :routing_key => nil,
        :block => false,
        :polling_interval => 1,
        :polling_timeout => nil
      }.update(opts)
      
      current_time = Time.new.to_f
      messages = []
      
      query = {
          '_meta.expired' => false,
          '_meta.visible_after' => {
            '$lt' => current_time
          }
      }
      if opts[:routing_key] != nil
        query['_meta.routing_key'] = opts[:routing_key]
      else
        query['_meta.routing_key'] = {
          '$exists' => false
        }
      end
      
      update = false
      if opts[:visibility_timeout] > -1.0
        update = {
            '$set' => {
              '_meta.visible_after' => current_time + opts[:visibility_timeout]
            }
        }
      end
      
      raw_messages = []
      
      # if we want to block, loop and sleep until messages are available (or we time out)
      if opts[:block]
        block_until_message_available( query, opts[:polling_interval], opts[:polling_timeout] )
      end

      if update
        (0..opts[:messages_read]).each do
          begin
            
            raw_message = @queue_collection.find_and_modify(:query => query, :update => update)
            
            if raw_message
              raw_messages << raw_message
            else
              break
            end
          
          rescue Mongo::OperationFailure => operation_failure
            if not  operation_failure.to_s.match(Queue::NO_OBJECT_FOUND_ERROR)
              raise operation_failure
            end
          end
          
        end
      else
        @queue_collection.find(query).limit(opts[:messages_read]).each do |raw_message|
          raw_messages << raw_message
        end
      end
      
      raw_messages.each do |raw_message|
        message = Karait::Message.new(raw_message=raw_message, queue_collection=@queue_collection)
        if not message.expired?
          messages << message
        end
      end
      
      return messages
    end
    
    def delete_messages(messages)
      ids = []
      messages.each {|message| ids << message._get_id}
      @queue_collection.update(
          {
              '_id' => {
                '$in' => ids
              }
          },
          {
              '$set' => {
                  '_meta.expired' => true
              }
          },
          :multi => true,
          :safe => true
      )
    end
    
    private
    
    def block_until_message_available(query, polling_interval=1.0, polling_timeout=None)
        current_time = Time.new.to_f
        
        while @queue_collection.find(query).count() == 0
            if polling_timeout and (Time.new.to_f - current_time) > polling_timeout
                break
            end
            
            sleep polling_interval
        end
    end
    
    def set_instance_variables(opts)
      
      defaults = {
        :host => 'localhost',
        :port => 27017,
        :database => 'karait',
        :queue => 'messages',
        :average_message_size => 8192,
        :queue_size => 4096
      }.merge(opts)
      
      @host = defaults[:host]
      @port = defaults[:port]
      @database = defaults[:database]
      @queue = defaults[:queue]
      @average_message_size = defaults[:average_message_size]
      @queue_size = defaults[:queue_size]
      
    end
    
    def create_mongo_connection
      @connection = Mongo::Connection.new(
        @host,
        @port
      )
      @database = @connection[@database]
      create_capped_collection
      @queue_collection = @database[@queue]
      @queue_collection.create_index('_id')
      @queue_collection.create_index('_meta.routing_key')
      @queue_collection.create_index('_meta.expired')
      @queue_collection.create_index('_meta.visible_after')
    end
    
    def create_capped_collection
      @database.create_collection(
        @queue,
        :size => (@average_message_size * @queue_size),
        :capped => true,
        :max => @queue_size
      )
    end
  end
end

File: node/lib/queue.js
var extend = require('./helpers').extend,
    Message = require('./message').Message,
    puts = require('sys').puts,
    mongodb = require('mongodb'),
    Db = mongodb.Db,
    Connection = mongodb.Connection,
    Server = mongodb.Server;
    
var nativeParserError = 'Native bson parser not compiled';
var makeFloat = 0.0000001;
    
exports.Queue = function(params, onQueueReady) {
    
    if (typeof(params) === 'function') {
        onQueueReady = params;
        params = {};
    }
    
    var defaults = {
        host: 'localhost',
        port: 27017,
        database: 'karait',
        queue: 'messages',
        averageMessageSize: 8192,
        queueSize: 4096
    };
    extend(this, defaults, params);
    this.onQueueReady = onQueueReady || function() {};
    this._initializeQueue(true);
};

exports.Queue.prototype._initializeQueue = function(nativeParser) {
    var _this = this;
    
    try {
        var db = new Db(
            this.database,
            new Server(this.host, this.port, {}),
            {native_parser: nativeParser}
        );
    } catch (e) {
        if (e.indexOf(nativeParserError) > -1) {
           this._initializeQueue(false);
           return;
        } else {
           this.onQueueReady(err, null);
           return;
        }
    }
    
    db.open(function(err, db) {
        if (err) {
            _this.onQueueReady(err, null);
            return;
        } else {
            _this.db = db;
            _this._createCappedCollection(db);
        }
    });
};

exports.Queue.prototype._createCappedCollection = function(db) {
    var _this = this;
    this.db.createCollection(
        this.queue,
        {
            'capped': true,
            'size': (this.averageMessageSize * this.queueSize),
            'max': this.queueSize,
        }, 
        function(err, collection) {
            if (err) {
                _this.onQueueReady(err, null);
                return;
            }
            _this.queueCollection = collection;
            _this._createIndexes();
            _this.onQueueReady(null, _this);
        }
    );
};

exports.Queue.prototype._createIndexes = function() {
    this.queueCollection.createIndex('_id', function(){});
    this.queueCollection.createIndex('_meta.routing_key', function() {});
    this.queueCollection.createIndex('_meta.expired', function() {});
    this.queueCollection.createIndex('_meta.visible_after', function(){});
}

exports.Queue.prototype.deleteMessages = function(messages, callback) {
    callback = callback || {};
    
    var deleteIds = [];
    for (var i = 0, message; (message = messages[i]) != null; i++) {
        deleteIds.push(message._source._id);
    }
        
    this.queueCollection.update(
        {
            _id: {
                $in: deleteIds
            }
        },
        {
            $set: {
                '_meta.expired': true
            }
        },
        {
            safe: true,
            multi: true
        },
        callback
    );
};

exports.Queue.prototype.write = function(message, params, callback) {
    
    if (typeof(params) === 'function') {
        callback = params;
        params = {};
    }
    
    callback = callback || function() {};
    
    if (message.toObject) {
        var messageObject = message.toObject();
    } else {
        var messageObject = message;
    }
    
    messageObject._meta = {
        expire: (params.expire || -1) - makeFloat,
        timestamp: (new Date()).getTime() / 1000.0,
        expired: false,
        visible_after: -1 - makeFloat
    }
    
    if (params.routingKey) {
        messageObject._meta.routing_key = params.routingKey;
    }
    
    this.queueCollection.insert(messageObject, {safe: true}, callback);
};

exports.Queue.prototype.read = function(params, callback) {
    var _this = this,
        params = params || {};
    
    if (typeof(params) === 'function') {
        callback = params;
        params = {};
    }
    
    callback = callback || function() {};
    
    extend(
        params,
        {
            messagesRead: 10,
            visibilityTimeout: -1.0 - makeFloat,
            routingKey: null
        },
        params
    );
    
    var currentTime = (new Date()).getTime() / 1000.0,
        query = {
            '_meta.expired': false,
            '_meta.visible_after': {
                '$lt': currentTime
            }
        },
        update = false;
    
    if (params.routingKey) {
        query['_meta.routing_key'] = params.routingKey
    } else {
        query['_meta.routing_key'] = {
            '$exists': false
        }
    }
    
    if (params.visibilityTimeout > -1.0) {
        update = {
            '$set': {
              '_meta.visible_after': currentTime + params.visibilityTimeout
            }
        }
    }
    
    if (!update) {
        this._normalFind(query, params.messagesRead, callback);
    } else {
        this._atomicFind(query, update, params.messagesRead, callback);
    }
};

exports.Queue.prototype._normalFind = function(query, limit, callback) {
    
    var messages = [],
        _this = this;
        
    this.queueCollection.find(query, {limit: limit}, function(err, cursor) {
        if (err) {
            callback(err, null);
            return;
        } else {
           cursor.toArray(function(err, items) {
               if (err) {
                   callback(err, null);
                   return;
               } else {
                   for (var i = 0, item; (item = items[i]) != null; i++) {
                       var message = new Message(item, _this.queueCollection);
                       if (!message.isExpired()) {
                           messages.push(message);
                       }
                   }
                   callback(null, messages);
               }
           });
        }
    });
};

exports.Queue.prototype._atomicFind = function(query, update, limit, callback) {
    var messages = [],
        count = 0,
        _this = this;
        
   (function fetchDocument() {
        _this.queueCollection.findAndModify(query, [], update, {}, function(err, item) {
            if (item) {
                var message = new Message(item, _this.queueCollection);
                
                if (!message.isExpired()) {
                    messages.push(message);
                }
            }
            
            if ( (count++) < (limit - 1) ) {
                fetchDocument();
            } else {
                callback(null, messages);
            }
        })
    })();
};